Post sealed box & Decrypt response

jongbonga opened the pull request:
Issue https://github.com/registreerocks/trusted-compute-web/issues/13

jongbonga committed:
more interfaces

jongbonga committed:
post messageData

jongbonga committed:
post and decrypt box

jongbonga committed:
secret key getter, sealbox action

jongbonga committed:
decrypt message

longtomjr commented on src/components/AccessForm.vue:
```diff
@@ -73,24 +79,66 @@ export default defineComponent({
         }
       );
+    },
+    async postEncryptedBox(messageData: MessageData) {
```
I think it makes sense to move this into vuex actions rather than doing the calls in the components (like we are doing with the file upload).

jongbonga replied:
I didn't put it in vuex since it's not a reusable call and the result isn't saved.
But it makes sense to centralize the http calls

longtomjr replied:
Now that I think about it. The result can be in the session state (since it will have to be in state somewhere if we display it), we should just clear it after the user closes the dialog or go to a different page.

I think having the encryption and decryption function calls along will all the server calls in vuex makes sense from a structure and auditability perspective.

jongbonga replied:
No need to save it if we clear it when the dialog box closes unless we allow it to be seen multiple times.
We can trigger the dialog on decryption success without store the result.

What's your thought?

longtomjr replied:
Well, if we do it through vuex actions, we will need to modify the state to be able to display the results, since, afaik, vuex actions should not return a result. Whether we store it inside of the state of the dialog (as in the component's state, not the app state), or the app state does not matter for this, as long as we clear the vuex state when the dialog closes.

I think it makes sense to be explicit about the state in this case. So do the call -> update values in vuex -> display dialog -> close dialog -> erase values from vuex.

PiDelport commented on src/components/AccessForm.vue:
```diff
@@ -73,24 +79,66 @@ export default defineComponent({
+          console.log(result);
+          if (result.status === 200) {
+            const { execution_token, nonce } = result.data;
+            await notifyErrors("Attestation failed", async () => {
+              const decryptedResult = decryptMessage(
+                execution_token,
+                nonce,
+                this.enclavePublicKey,
+                this.ourSecretKey
+              );
+              console.log(decryptedResult);
```
(Using `notifyErrors` here has no effect, since `decryptMessage` doesn't raise exceptions.)

PiDelport commented on src/components/AccessForm.vue:
```diff
@@ -73,24 +79,66 @@ export default defineComponent({
+      };
+      await this.axios
+        .post("https://rtc-data.registree.io/auth/tokens", payload)
+        .then(async result => {
```
We should probably avoid mixing both async/await and `Promise` syntax?

Standardising on async/await probably makes most sense here, for better tracebacks?

PiDelport commented on src/components/AccessForm.vue:
```diff
@@ -73,24 +79,66 @@ export default defineComponent({
+          }
+        })
+        .catch(error => {
+          console.log(`post resul ${error}`);
```
(typo)

jongbonga committed:
moved post encryted box method to vuex

jongbonga committed:
save decrypted msg

jongbonga committed:
update mutation nomenclature

jongbonga committed:
show msg dialog and reset on close

longtomjr commented on src/components/AccessForm.vue:
```diff
@@ -73,24 +78,48 @@ export default defineComponent({
         async (valid: boolean) => {
           if (valid) {
-            const enclavePublicKey = this.$store.getters.enclavePublicKey;
+            this.loading = true;
```
I might be missing it, but is this set to false after the calls?

jongbonga replied:
Oh yes, since I call setting it to false on `finally`. 
I forgot to update it after changing to async await.

longtomjr commented on src/store/index.ts:
```diff
@@ -88,33 +102,70 @@ export default createStore<State>({
         return "error";
       }
-      await dispatch("decryptUploadResponse", res.data);
+      const msg = await dispatch("decryptResponse", res.data);
```
Afaik vuex results should not return values that is used. For this is probably makes more sense to just make `decryptResponse` a normal function, and call it from the relevant places, rather than an action, and pass in all the data it needs from the state as parameters.

jongbonga replied:
Why shouldn't we return results from Actions?

longtomjr replied:
Vuex follows a "one-way data flow" data flow model. From my understanding, once you start returning values from actions, you break the one-way flow model. So the goal is to separate actions on data from access to the data.

longtomjr commented on src/store/index.ts:
```diff
@@ -88,33 +102,70 @@ export default createStore<State>({
       const uuid = message.slice(24);
       commit("saveUploadResult", { accessKey, uuid });
+    },
+    async sealBox({ commit, dispatch, getters }, data) {
```
I think it makes more sense to call this action something like `requestExecutionToken`

longtomjr reviewed (changes requested):
For the vuex actions. It makes sense to think of them as a chain of tasks that needs to be executed. This helps when it comes to naming, and it makes it easier to reason about them. Another thing is that the action that will be called from a component should describe the whole action chain.

For this action chain I can imagine something like this:
Name of the action chain (so the first action name): `provisionExecutionToken`
and the chain:
```
provisionExecutionToken(does encryption bit)
-> requestExecutionToken(does the server call bit, and then calls the js function to decrypt response)
-> setExecutionToken (mutation that sets the state)
```
It does not need to follow this exactly, but just wanted to give an example of what I am talking about.

jongbonga committed:
stop loading after box is sealed and posted

jongbonga committed:
action nomanclature

jongbonga committed:
no return from action

longtomjr commented on src/store/index.ts:
```diff
@@ -80,41 +94,98 @@ export default createStore<State>({
+      const { nonce, ciphertext } = res.data;
       const enclavePubKey = getters.enclavePublicKey;
-      if (!enclavePubKey || !state.ourSecretKey) {
+      const ourSecretKey = getters.ourSecretKey;
```
Not sure I am following why a getter is used here instead of using the state. Am I missing something?

longtomjr commented on src/store/index.ts:
```diff
@@ -80,41 +94,98 @@ export default createStore<State>({
+    async requestExecutionToken({ commit, dispatch, getters }, data) {
+      const enclavePubKey = getters.enclavePublicKey;
+      if (!enclavePubKey) {
+        return null;
```
We might want to error out here rather than just returning. If this is temporary until the endpoints are up, just add a `// TODO`

PiDelport replied:
(This still needs resolving, right? Implementing the other suggestion to make the `enclavePublicKey` getter throw an error on `null` will also resolve this, though.)

jongbonga committed:
use ourSecretKey's state instead of getters

jongbonga committed:
// TODO: Error Handling

longtomjr commented on src/components/AccessForm.vue:
```diff
@@ -73,24 +78,49 @@ export default defineComponent({
             // Get the unproxied form data before encrypting, for clarity.
             const data = Object.assign({}, this.form);
-            await encryptJson(data, enclavePublicKey);
+            await this.requestExecutionToken(data);
```
We should double check what happens if this action fails, since we are not wrapping it in try-catch block.

longtomjr commented on src/store/index.ts:
```diff
@@ -80,41 +90,101 @@ export default createStore<State>({
+        ciphertext,
+        nonce,
+        enclavePubKey,
+        ourSecretKey as Base64
```
We should check that the variable exists instead of casting it.

longtomjr commented on src/store/index.ts:
```diff
@@ -80,41 +90,101 @@ export default createStore<State>({
+        nonce,
         enclavePubKey,
-        state.ourSecretKey
+        ourSecretKey as Base64
```
Same as above, this cast should not be necessary if we check for null.

We should check what to do with the different cases where the `ourSecretKey` is `null`. In this case it seems like a clear exception, something that should not be possible. In other cases we might just want to restart with the attestation process.

jongbonga committed:
check if ourSecretKey is null

jongbonga committed:
catch requestExecutionToken

PiDelport commented on src/store/index.ts:
```diff
@@ -80,41 +90,103 @@ export default createStore<State>({
         request
       );
-      if (res.status !== 200) {
+      const enclavePubKey = getters.enclavePublicKey;
+      const ourSecretKey = state.ourSecretKey;
```
Rather than checking these for `null` at every usage site, we should probably change these getters to do the check and throw an error for `null`, for now? (In future, they could lazily attest, as needed?)

jongbonga replied:
we can discuss this tomorrow, @longtomjr said there was no need for a getter to retrieve `ourSecretKey`

PiDelport replied:
Cool, I added a `checkDefined` helper for this.

PiDelport commented on src/store/index.ts:
```diff
@@ -80,41 +90,103 @@ export default createStore<State>({
+      const enclavePubKey = getters.enclavePublicKey;
+      const ourSecretKey = state.ourSecretKey;
+
+      if (res.status !== 200 || !enclavePubKey || !ourSecretKey) {
```
We probably want to replace this status check with an option passed to the Axios `post` call:

```js
{ validateStatus: status => status === 200 }
```

jongbonga replied:
Did we agree on the `200`?

PiDelport replied:
I think so, yeah: anything other than 200 should be unexpected by this code.

PiDelport commented on src/store/index.ts:
```diff
@@ -80,41 +90,103 @@ export default createStore<State>({
+
+      if (res.status !== 200 || !enclavePubKey || !ourSecretKey) {
+        // TODO: Error Handling
+        return "error";
```
We probably want to make all these placeholder error cases use `throw`, rather than `return`?

PiDelport commented on src/utils/cryptography.ts:
```diff
@@ -6,15 +6,19 @@ import util from "tweetnacl-util";
+
+export interface OurData {
+  ourSecretKey: Base64;
+}
+
 export interface EncryptedMessage {
-  readonly ourData: {
-    ourSecretKey: Base64;
-  };
-  readonly messageData: {
-    ourPublicKey: Base64;
-    nonce: Base64;
-    ciphertext: Base64;
-  };
+  readonly ourData: OurData;
+  readonly messageData: MessageData;
```
Do we still want to pull these out, or should we rather roll this back?

My intuition is that we probably don't want to name or expose these sub-parts of `EncryptedMessage` too much: we'll probably end up refactoring this once we have better secret key management in place?

jongbonga replied:
I'm not seeing an issue with them being decoupled since `EncryptedMessage` exposes them anyway

PiDelport replied:
Discussed in call: I pulled these back in, and added a TODO comment about the key management.

PiDelport commented on src/store/index.ts:
```diff
@@ -14,6 +19,7 @@ export interface State {
   ourSecretKey: Base64 | null;
   attestationResult: AttestationToken | null;
   uploadResult: { accessKey: string; uuid: Uint8Array } | null;
+  decryptedMessage: string | null;
```
`executionToken` is probably a more meaningful name for this and its references? We should also probably keep it as a `Uint8Array` for now: that's what the code will actually store here, currently.

(We might later considering storing it specifically as a JWT, if we settle on that from the server side, but that's not in place yet.)

```suggestion
  executionToken: Uint8Array | null;
```

PiDelport commented on src/store/index.ts:
```diff
@@ -50,6 +57,9 @@ export default createStore<State>({
     ) {
       state.uploadResult = uploadResult;
+    },
+    setDecryptedMsg(state, msg) {
```
(As above, with some type signatures)

```suggestion
    setExecutionToken(state: State, executionToken: Uint8Array) {
```

PiDelport commented on src/components/AccessForm.vue:
```diff
@@ -73,24 +78,54 @@ export default defineComponent({
+    async handleMsgDisplay(message: string) {
+      await ElMessageBox({
+        title: "Something",
+        message,
```
We should update this to take a `executionToken: Uint8Array` (rather than `message: string`).

This function should also do some basic formatting of the binary data for user display, for now, even if we just Base64-encode it: something like this?

```ts
`Please save this execution token: ${encodeBase64(executionToken)}`
```

PiDelport commented on src/store/index.ts:
```diff
@@ -80,41 +90,103 @@ export default createStore<State>({
-      response: UploadResponse
-    ) {
+    async postAccessForm({ commit, getters, state }, request: UploadRequest) {
+      const res = await axios.post<UploadResponse>(
```
It probably doesn't make sense to reuse the data upload `UploadRequest` / `UploadResponse` here… we should either use separate types for the data upload and execution token interactions' requests and responses, or we should keep the models more general.

My hunch is it's probably easiest to do the latter, for now: we could make the current models `SealedRequest` / `SealedResponse` instead?

PiDelport replied:
There's also a lot of duplicated code between `uploadFile` and `postAccessForm` right now: we should probably factor that out into a small helper, for now?

Straw suggestion: something like `sealedPost`, with some parameters for what differs, and which handles all the common sealing and unsealing logic

PiDelport replied:
Cool, I renamed the types to use "sealed" as terminology, for now.

PiDelport replied:
I factored out a `sealedPost` helper: d1329227

PiDelport commented on src/store/index.ts:
```diff
@@ -80,41 +90,103 @@ export default createStore<State>({
+      );
+
+      commit("saveSecretKey", ourData.ourSecretKey);
+      dispatch("postAccessForm", {
```
```suggestion
      await dispatch("postAccessForm", {
```

(The "loading" state of the execution token request / submit button awaits this.)

PiDelport commented on src/store/index.ts:
```diff
@@ -80,41 +90,103 @@ export default createStore<State>({
       const uuid = message.slice(24);
       commit("saveUploadResult", { accessKey, uuid });
+    },
+    async requestExecutionToken({ commit, dispatch, getters }, data) {
```
We should probably start documenting and typing some of the main actions, for readability? How about:

```suggestion

    /**
     * Submit an execution token request.
     *
     * @see #actions#postAccessForm
     * @see #mutations#setExecutionToken
     */
    async requestExecutionToken(
      { commit, dispatch, getters },
      executionTokenRequest
    ): Promise<void> {
```

PiDelport commented on src/store/index.ts:
```diff
@@ -80,41 +90,103 @@ export default createStore<State>({
+        enclavePubKey,
+        state.ourSecretKey
+      );
     },
```
Dead code?

```suggestion
```

PiDelport replied:
(Dropped)

PiDelport commented on src/components/AccessForm.vue:
```diff
@@ -30,17 +30,22 @@
+        :loading="loading"
+        @click="onSubmit('accessForm')"
       >
+        Create
```
"Request" probably makes more sense, since this is requesting a new execution token?

```suggestion
        Request
```

PiDelport commented on src/store/index.ts:
```diff
@@ -80,41 +90,103 @@ export default createStore<State>({
       }
-      await dispatch("parseUploadMessage", msg);
+
+      //TODO: Open Dialog Box show message
```
Is this still relevant, with the `watch` logic in the `AccessForm` component?

jongbonga replied:
No, updating now. thanks.

PiDelport commented on src/components/AccessForm.vue:
```diff
@@ -73,24 +78,54 @@ export default defineComponent({
             // Get the unproxied form data before encrypting, for clarity.
             const data = Object.assign({}, this.form);
-            await encryptJson(data, enclavePublicKey);
+            try {
+              await this.requestExecutionToken(data);
+            } catch (e) {
+              console.log(e);
+              // TODO: Error Handling
+            }
```
We could use the `notifyErrors` wrapper here?

```suggestion
            await notifyErrors("Execution token request failed", () => {
              return this.requestExecutionToken(data);
            });
```

PiDelport commented on src/store/index.ts:
```diff
@@ -80,41 +90,103 @@ export default createStore<State>({
-      { state, dispatch, getters },
-      response: UploadResponse
-    ) {
+    async postAccessForm({ commit, getters, state }, request: UploadRequest) {
```
It's probably worth making all the action return types explicit, to avoid accidental returns.

```suggestion
    async postAccessForm(
      { commit, getters, state },
      request: UploadRequest
    ): Promise<void> {
```

PiDelport commented on src/store/index.ts:
```diff
@@ -80,41 +90,103 @@ export default createStore<State>({
       });
     },
-    async uploadFile({ dispatch }, request: UploadRequest) {
+    async uploadFile({ dispatch, getters, state }, request: UploadRequest) {
```
It's probably worth making all the action return types explicit, to avoid accidental returns.

```suggestion
    async uploadFile(
      { dispatch, getters, state },
      request: UploadRequest
    ): Promise<void> {
```

PiDelport commented:
I know this is quite a lot of feedback: let me know if you'd like me to tackle some of it, or have a session to go through it?

jongbonga committed:
refactor(store): making all the action return types explicit

Co-authored-by: Pi Delport <pi@registree.io>

jongbonga committed:
refactor(AccessForm): use notifyErrors around requestExecutionToken

Co-authored-by: Pi Delport <pi@registree.io>

jongbonga committed:
refactor(AccessForm): change create to request

Co-authored-by: Pi Delport <pi@registree.io>

jongbonga committed:
throw instead of returning error

jongbonga committed:
change decryptedMessage to executionToken

PiDelport committed:
chore(AccessForm): remove TODO

Co-authored-by: Jonathan Ngbonga <jongbonga@gmail.com>

PiDelport commented:
(Rebased to refresh and update)

PiDelport committed:
refactor(store): use axios validateStatus to check the status code

This will give us a bit more HTTP debugging context on failure, compared
to checking and throwing later.

PiDelport committed:
refactor(store,AccessForm): rename setDecryptedMsg → setExecutionToken, and use Uint8Array

PiDelport committed:
text(AccessForm): tweak execution token message box title

PiDelport committed:
docs(store): add note about execution token format

PiDelport committed:
refactor(store): tweak requestExecutionToken, throw on error

PiDelport committed:
fix(store): add await for postAccessForm dispatch

PiDelport committed:
lint(cryptography): yarn lint

PiDelport committed:
refactor(cryptography): pull fields back into EncryptedMessage

PiDelport committed:
docs(cryptography): add TODO comment for better local key management

PiDelport committed:
refactor(store): drop unused code: decryptResponse

PiDelport committed:
refactor(models): rename data upload types to "sealed" types

* @/models/data-upload → @/models/sealed-calls
* UploadRequest → SealedRequest
* UploadResponse → SealedResponse

PiDelport committed:
feat(utils): add checkDefined helper

PiDelport committed:
refactor(store): use checkDefined for null checks

Also add some type annotations, while here.

PiDelport commented on src/components/AccessForm.vue:
```diff
@@ -73,24 +80,50 @@ export default defineComponent({
     };
   },
+  computed: {
+    ...mapState(["ourSecretKey", "executionToken"]),
+    ...mapGetters(["enclavePublicKey"])
```
We can remove the unused accessors here.

```suggestion
    ...mapState(["executionToken"])
```

PiDelport committed:
refactor(AccessForm): remove unused store accessors

PiDelport committed:
refactor(store): pull out UploadResult type

PiDelport committed:
refactor(store): factor out a common sealedPost helper

PiDelport committed:
refactor(store): drop unused secret key from store state

Each request's ephemeral secret key is now managed inside sealedPost.

PiDelport reviewed (approved):
@jongbonga and I had a session together to resolve the remaining conversations, with @Nghondzweni listening in. Thanks! 😊️

After that, I made to factor out a `sealedPost` helper substantially simplifies the action code: both `encryptAndUploadFile` and `requestExecutionToken` are now just small wrappers around it. @jongbonga and @longtomjr, can you review this change before merging?

longtomjr commented on src/store/index.ts:
```diff
@@ -154,3 +136,56 @@ async function fetchAttestationToken(): Promise<string> {
+ * @returns - The unsealed response bytes
+ */
+async function sealedPost<T>(
+  context: ActionContext<State, State>,
```
I would prefer passing in the public key value instead of action context.

PiDelport replied:
Yeah, that's a good idea, after talking through it: thanks! I'll push an update.

PiDelport replied:
Updated: [aa891f6..7c17610](https://github.com/registreerocks/trusted-compute-web/compare/aa891f6..7c17610)

longtomjr reviewed (approved):
Looks good. Thanks everyone for getting this sorted!
